addon/mixins/stateful.js

```diff
@@ -0,0 +1,299 @@
+import Ember from 'ember';
+import { task, taskGroup } from 'ember-concurrency';
+import ERRORS from 'ember-stateful/errors';
+
+const { computed } = Ember;
+
+function isThenable(obj) {
+  return obj && typeof(obj.then) === 'function';
+}
+
+function executeStateHook(target, hook) {
```

Comment:
this being a generator function instead will let us process functions, functions returning a promise, as well as generators directly. This way the state hook could be defined in the following way:
```js
...
  A: {
    *_try() {
      yield somePromise;
      if (yield anotherPromise) {
        this.transitionTo('someOtherState');
      }
    }
  }
```

Reply:
I am not sure that plain generator will work in this case, from what I tested plain generators can't yield on Promises, that's why I did not do it with generators. But this will be ultimately (through `yield*`) yielded from a Task so maybe it will work. I will test it out.